example_buffer.py

```diff
@@ -78,24 +82,26 @@ def update(self, new_games, samples_per_game=4):
                 continue
             elif first_new_game is None:
                 first_new_game = idx
-                print(
-                    "Found {}/{} new games".format(len(new_games) - idx, len(new_games)))
-
-            choices = [(timestamp, ex) for ex in pick_examples_from_tfrecord(
-                game, samples_per_game)]
-            self.examples.extend(choices)
+                print("Found {}/{} new games".format(
+                    len(new_games) - idx, len(new_games)))
+            self.examples.extend(self.func(game))
 
     def flush(self, path):
+        # random.shuffle on deque is O(n^2) convert to list for O(n)
+        self.examples = list(self.examples)
+        random.shuffle(self.examples)
```

Comment:
https://bugs.python.org/issue4123
Nice catch. I had always assumed deques were just implemented as arrays with overallocated space on both ends, rather than as a linked list.